### ISSUE TITLE
negative std?

### ISSUE THREAD

**Issue by SiyuanYan1** [ml-research/NeSyXIL#1]
In both the pre-trained model and model, the std is negative, which is not valid to train.

**Comment by wolfstam** [comment]
Hey thanks so much for letting us know. I am assuming you are referring to this line in the slot attention module, right?

        self.slots_log_sigma = nn.Parameter(torch.randn(1, 1, dim))
(https://github.com/ml-research/NeSyConceptLearner/blob/2f3fea4e7898280ae4105bf037282fd6a60c1ef1/src/model.py#L19)

Otherwise could you specify more precisely what std you are referring to?

**Comment by SiyuanYan1** [comment]
Yes, if I directly train the model, I will get an error that std should be >=0. After I change std to 0, I only can get 40% test accuracy.

**Comment by wolfstam** [comment]
Ok this is probably related to the following issue: https://discuss.pytorch.org/t/how-does-torch-normal-work-with-a-negative-standard-deviation/107571
The newer pytorch versions had therefore removed the possibility of passing negative sigma values. The code should work with the attached docker file. 
Otherwise, could you please try and replace the line mentioned above with the following (you might have to change the device if you are not using cuda) and let me know if that works then I can fix this in the repo:

        self.slots_log_sigma = nn.Parameter(torch.randn(1, 1, dim)).abs().to(device='cuda')

**Comment by SiyuanYan1** [comment]
Thank you so much! It works now.

**Comment by wolfstam** [comment]
Glad it worked thanks for the feedback!
